### ISSUE TITLE
object is possibly undefined for imported object

### ISSUE THREAD

**Issue by mjbvz** [microsoft/TypeScript#13369]
_From @unional on January 9, 2017 19:57_

- VSCode Version: 1.8.1

```ts
// a.ts

export interface Boo {
   z: string
}

export interface Foo {
  x: string
  y?: Boo
}

const value: Foo = { x: '1' }
const { x, y } = value
export { x, y } // yes, `y` can be undefined

// b.ts
import { x, y } from './a'

function do() {
  if (y) {
    console.log(y.z) // object is possibly 'undefined'
  }
}
```

It seems like `tsc` cannot determine the type of `y` being `Boo | undefined`. The code completion doesn't show either.

_Copied from original issue: Microsoft/vscode#18339_

**Comment by mjbvz** [comment]
This does not look VSCode specific. I'm also able to repo it using TS 2.1.4 and the latest 2.2 with the following `tsconfig.json`:

```json
{
    "compilerOptions": {
        "strictNullChecks": true
    }
}
```

**Comment by ericmuyser** [comment]
Is there/can there be some syntax for things that can be undefined? I have this:

`categories?: Array<ProductCategory> = undefined`

Usage:
`product.categories = await Promise.all(product.categories.map(async (category) => await this.SaveCategory(category)))`

Could usage be `product.categories?.map`?

I'm expecting at some point it gets set to an array (hydrating from server). The reason is I have other stuff like `shipping_taxable?: string = undefined`. I don't want that to be `shipping_taxable: string = ''`. Or `id: number = 0`. If it's not hydrated and happens to be undefined, I want it to error out, not have an irrelevant default value. It's also useful for checking what changed.

**Comment by RyanCavanaugh** [comment]
You can use `product.categories!.map(...` to indicate that you believe the value to be non-null/undefined

**Comment by RyanCavanaugh** [comment]
Oh, and the other thing you can do is write
```ts
categories: Array<ProductCategory> = undefined!;
```
so that you don't need `!` at the use sites. Sort of a trade-off. I would do this method if the field were initialized "early" in the container's lifetime, and the other approach if it were more common to find the field uninitialized at a later point.

**Comment by jakebailey** [comment]
Just checking out old issues; this one was fixed by #16658.
